stix2/equivalence/object_equivalence/__init__.py

```diff
@@ -0,0 +1,434 @@
+                    url_match = True
+
+            # Special case: if source_name is a STIX defined name and either
+            # external_id or url match then its a perfect match and other entries
+            # can be ignored.
+            if sn_match and (ei_match or url_match) and source_name in allowed:
+                result = 1.0
+                logger.debug(
+                    "--\t\tpartial_external_reference_based '%s' '%s'\tresult: '%s'",
+                    refs1, refs2, result,
+                )
+                return result
+
+            # Regular check. If the source_name (not STIX-defined) or external_id or
+            # url matches then we consider the entry a match.
+            if (sn_match or ei_match or url_match) and source_name not in allowed:
+                matches += 1
+
+    result = matches / max(len(refs1), len(refs2))
+    logger.debug(
+        "--\t\tpartial_external_reference_based '%s' '%s'\tresult: '%s'",
+        refs1, refs2, result,
+    )
+    return result
+
+
+def partial_location_distance(lat1, long1, lat2, long2, threshold):
+    """Given two coordinates perform a matching based on its distance using the Haversine Formula.
+
+    Args:
+        lat1: Latitude value for first coordinate point.
+        lat2: Latitude value for second coordinate point.
+        long1: Longitude value for first coordinate point.
+        long2: Longitude value for second coordinate point.
+        threshold (float): A kilometer measurement for the threshold distance between these two points.
+
+    Returns:
+        float: Number between 0.0 and 1.0 depending on match.
+
+    """
+    from haversine import Unit, haversine
+    distance = haversine((lat1, long1), (lat2, long2), unit=Unit.KILOMETERS)
+    result = 1 - (distance / threshold)
+    logger.debug(
+        "--\t\tpartial_location_distance '%s' '%s' threshold: '%s'\tresult: '%s'",
+        (lat1, long1), (lat2, long2), threshold, result,
+    )
+    return result
+
+
+def _versioned_checks(ref1, ref2, ds1, ds2, **weights):
+    """Checks multiple object versions if present in graph.
+    Maximizes for the semantic equivalence score of a particular version."""
+    results = {}
+    objects1 = ds1.query([Filter("id", "=", ref1)])
+    objects2 = ds2.query([Filter("id", "=", ref2)])
+
+    if len(objects1) > 0 and len(objects2) > 0:
+        for o1 in objects1:
+            for o2 in objects2:
+                result = semantically_equivalent(o1, o2, **weights)
+                if ref1 not in results:
+                    results[ref1] = {"matched": ref2, "value": result}
+                elif result > results[ref1]["value"]:
+                    results[ref1] = {"matched": ref2, "value": result}
+    result = results.get(ref1, {}).get("value", 0.0)
+    logger.debug(
+        "--\t\t_versioned_checks '%s' '%s'\tresult: '%s'",
+        ref1, ref2, result,
+    )
+    return result
+
+
+def semantic_check(ref1, ref2, ds1, ds2, **weights):
```

Comment:
It seems like this function is checking de-referenced objects rather than semantics in general. Maybe rename to `reference_check` or `dereferenced_check`? Same with `list_semantic_check`.